pgx-pg-sys/src/submodules/panic.rs

```diff
@@ -0,0 +1,423 @@
+        CaughtError::RustPanic {
+            ereport: ErrorReportWithLevel {
+                level: PgLogLevel::ERROR,
+                ereport: ErrorReport::with_location(
+                    PgSqlErrorCode::ERRCODE_INTERNAL_ERROR,
+                    "Box<Any>",
+                    take_panic_location(),
+                ),
+                stack: Default::default(),
+            },
+            payload: e,
+        }
+    }
+}
+
+fn do_ereport(ereport: ManuallyDrop<ErrorReportWithLevel>) {
+    // we define this here to make it difficult for not only pgx, but pgx users
+    // to find and directly call this function.  They'd have to do the same as
+    // this, and that seems like more work than a normal programmer would want to do
+    extern "C" {
+        fn pgx_ereport(
+            level: i32,
+            code: i32,
+            message: *const std::os::raw::c_char,
+            detail: *const std::os::raw::c_char,
+            funcname: *const std::os::raw::c_char,
+            file: *const std::os::raw::c_char,
+            lineno: i32,
+            contexts: *const std::os::raw::c_char,
+        );
+    }
+
+    unsafe {
+        // because of the calls to `.as_pg_cstr()`, which allocate using `palloc0()`,
+        // we need to be in the `ErrorContext` when we allocate those
+        //
+        // specifically, the problem here is `self.panic.location.file & .funcname`.  At the C level,
+        // Postgres expects these to be static strings, created at compile time, rather
+        // than something allocated from a MemoryContext.  Our version of ereport (pgx_ereport)
+        // accepts a user-provided strings for them, so we can report function/file/line information
+        // from rust code.
+        //
+        // We just go ahead and allocate all the strings we need in the `ErrorContext` for convenience
+        let old_cxt = MemoryContextSwitchTo(crate::ErrorContext);
+        let level = ereport.level as i32;
+        let errocode = ereport.ereport.errcode as i32;
+        let contexts = contexts_as_pg_cstr(&ereport.stack);
+        let funcname = ereport.ereport.funcname.as_ref().as_pg_cstr();
+        let file = ereport.ereport.location.file.as_str().as_pg_cstr();
+        let message = ereport.ereport.message.as_str().as_pg_cstr();
+        let detail = ereport.ereport.detail.as_ref().as_pg_cstr();
+        let line = ereport.ereport.location.line as i32;
+        MemoryContextSwitchTo(old_cxt);
+
+        // before calling `pgx_ereport` it's imperative we drop everything Rust-allocated we possibly can.
+        // `pgx_ereport` very well might `longjmp` to somewhere else, either in pgx or Postgres, and
+        // we'd rather not be leaking memory during error handling
+        //
+        // the few `.as_pg_cstr()`s do their allocation in Postgres' `CurrentMemoryContext`, so they'll
+        // be cleaned up by Postgres at the right time
+        let _ = ManuallyDrop::into_inner(ereport);
```

Comment:
I'm pretty sure this needs to use [`ManuallyDrop::drop(&mut ereport)`](https://doc.rust-lang.org/stable/std/mem/struct.ManuallyDrop.html#method.drop) instead. IIUC, having the drop automatically run in this function causes it no longer be a "plain old frame".

Reply:
I read the docs and they said if you have ownership of the thing then to use `::into_inner`.

I'll defer to you tho on that...

Reply:
It's not needed, you're right. Actually, ManuallyDrop is not -- it turns out that it is allowed for us to jump over stack frames with landing pads, so long as those landing pads are no-ops. This is surprising since it defeats the main benefit of making that UB, but honestly its *extremely* good news.